### DISCUSSION TITLE
Wiki update

### DISCUSSION THREAD

**Discussion by yeyeto2788** [tiny-pilot/tinypilot#893] [General]
Hey guys! 

Awesome work done on the project 🥳 . I just finished setting up my first device with tinypilot and looks promising 😃 .

I wanted to contribute with some minor issues I found on the documentation. Before I list them down let me first tell you about my setup.

**SBC:** Raspberry Pi Zero W
**HDMI capture:** C779 CSI-HDMI bridge module

Steps I followed:

1. Install everything as default following repo `README.md` (everything went smooth) you can see the output if needed on the [installation_001.log](https://github.com/tiny-pilot/tinypilot/files/7812848/installation_001.log) file.
2. Edit the `settings.yml` with the needed attributes/variables for the rpi 0 (**It failed**) the output is in the [installation_002.log](https://github.com/tiny-pilot/tinypilot/files/7812862/installation_002.log) file.
  * The attributes/variables were as in the [Pi Zero Optimizations](https://github.com/tiny-pilot/tinypilot/wiki/Pi-Zero-Optimizations) page:
    ```yaml
     ustreamer_desired_fps: 15
     ustreamer_drop_same_frames: 15
     ```
3. Edit once again the `settings.yml`with the `ustreamer_use_dv_timings: true` attribute/variable as I saw it on the issue [#93](https://github.com/tiny-pilot/tinypilot/issues/93#issuecomment-724285252) ending with a `setting.yaml` file like this:
   ```yaml
   ustreamer_capture_device: "tc358743"
   ustreamer_use_dv_timings: true
   ustreamer_desired_fps: 15
   ustreamer_drop_same_frames: 15
   ustreamer_port: 8001
   ustreamer_persistent: true
   ustreamer_encoder: omx
   ustreamer_format: uyvy
   ustreamer_resolution: 1920x720
   ```

Now to the files I visited that IMHO, have missing or wrong information:
- [Pi Zero Optimizations](https://github.com/tiny-pilot/tinypilot/wiki/Pi-Zero-Optimizations)
  - Would be great to add how to edit the `settings.yml` via command line as in other pages.
- [Example: TC358743 HDMI to CSI capture chip](https://github.com/tiny-pilot/tinypilot/wiki/Installation-Options#example-tc358743-hdmi-to-csi-capture-chip)
  - No reference to the `USE_TC358743_DEFAULTS` variable seen in the [quick-install](https://github.com/tiny-pilot/tinypilot/blob/master/quick-install#L57) script. (I know it's being set by reading the `settings.yaml` file but for some reason on the 3rd tried it work setting that variable before executing the script (Even with the script setting it up as empty on that line).
- [Installation Options](https://github.com/tiny-pilot/tinypilot/wiki/Installation-Options)
  - It references a `TINYPILOT_SETTINGS` variable which seems not correct as per the [quick-install](https://github.com/tiny-pilot/tinypilot/blob/master/quick-install#L42) looks for a **`TINYPILOT_SETTINGS_FILE`**
- [All pages above]()
  - I didn't find any information on how to create the user so I finally prior executing any of the commands proposed I needed to create the user and the group with
    ```console
    sudo useradd --create-home --system --user-group tinypilot
    ```
  - Maybe would be great to have a separate wiki page for editing the `settings.yml` file
- [Contributing](https://github.com/tiny-pilot/tinypilot/blob/master/CONTRIBUTING.md)
  - No section on how to edit the wiki or how to proceed or how to help on documenting.
  - No section regarding translation (I can help on the Spanish part)

P.D: 
- If any more information is needed, do ask me 😛 
- I can add these changes to the wiki if required. (Just need to know the procedure, like, if it is needed to activate the hooks when there is no code developed)

**Comment by mtlynch** [comment]
Thanks for this!

Sorry for the delay on responding. I'm hoping to get to this next week.

**Comment by yeyeto2788** [comment]
No problem. If you want me to do something, let me know.

**Comment by mtlynch** [comment]
Sorry again for the delay!

>Would be great to add how to edit the settings.yml via command line as in other pages.

We could, but it's a little ugly to make sure we're doing it idempotently. If we run the same command 10 times, it shouldn't just add the same line, but it's a bit more complicated to implement it to prevent that.

>No reference to the USE_TC358743_DEFAULTS variable seen in the quick-install script. (I know it's being set by reading the settings.yaml file but for some reason on the 3rd tried it work setting that variable before executing the script (Even with the script setting it up as empty on that line).

The user can't set the `USE_TC358743_DEFAULTS ` variable, as it gets [set in the script itself](https://github.com/tiny-pilot/tinypilot/blob/f415705497e3ae5d785821c69aa739902f759a7c/quick-install#L57). What you can do is set add this line to `/home/tinypilot/settings.yml`:

```yaml
ustreamer_capture_device: tc358743
```

It didn't take effect before because the documentation had quotes around `tc358743`, but the grep is [not flexible on quotes](https://github.com/tiny-pilot/tinypilot/blob/f415705497e3ae5d785821c69aa739902f759a7c/quick-install#L59). I've updated the examples to get rid of the quotes.

>It references a TINYPILOT_SETTINGS variable which seems not correct as per the quick-install looks for a TINYPILOT_SETTINGS_FILE

The two variables are independent. Both only have meaning locally within the script/example.

>I didn't find any information on how to create the user so I finally prior executing any of the commands proposed I needed to create the user and the group with

Ah, you're right. That should have been a first step in the advanced installation instructions. I've added that to the [examples](https://github.com/tiny-pilot/tinypilot/wiki/Installation-Options).

>No section on how to edit the wiki or how to proceed or how to help on documenting.

Github doesn't have tools for letting us review wiki changes effectively, so it's basically just that anyone's allowed to edit, and everyone does their best to keep it accurate and clear.

>No section regarding translation (I can help on the Spanish part)

We unfortunately don't have resources to support multi-language documentation or UI, but if you publish a Spanish-language tutorial, I'll link to it.

**Reply by yeyeto2788** [reply]
Wow!!! It does not matter the time it took you, as long as, all answers are like that one. BRAVO 👏🏽  I really love this kind of replies and I'm sorry for making probably the issue too long with too many things.

>It didn't take effect before because the documentation had quotes around tc358743, but the grep is not flexible on quotes. I've updated the examples to get rid of the quotes.

I'll try then this week (or next one) to double check this is not happening anymore.

>Ah, you're right. That should have been a first step in the advanced installation instructions. I've added that to the examples.

Awesome, I saw it on the command and it shouldn't fail anymore. 

>We unfortunately don't have resources to support multi-language documentation or UI

I was thinking more on a solution for translations like [this one](https://flask-babel.tkte.ch/) (latest change from Nov 2021 so it seems to be maintained > [repo](https://github.com/python-babel/flask-babel))but I do have to say that I've never tried it before but I'm pretty sure that more people like me are willing to contribute in their native languages and starting to translate doesn't mean you need to make the UI multi-language available as soon as people start contributing. This will probably take longer than translating itself.

**Comment by mtlynch** [comment]
>I was thinking more on a solution for translations like this one (latest change from Nov 2021 so it seems to be maintained > repo)but I do have to say that I've never tried it before but I'm pretty sure that more people like me are willing to contribute in their native languages and starting to translate doesn't mean you need to make the UI multi-language available as soon as people start contributing. This will probably take longer than translating itself.

There are definitely tools that we could use to support translation. That part is not too hard. The hard part is rearchitecting our app so that all UI strings go through a translation layer. Like [this error message](https://github.com/tiny-pilot/tinypilot/blob/f415705497e3ae5d785821c69aa739902f759a7c/app/templates/custom-elements/video-settings-dialog.html#L198):

```javascript
this.dispatchEvent(
  new DialogFailedEvent({
    title: "Failed to Load Video JPEG Quality",
    details: error,
  })
);
```

I think would have to become:

```javascript
this.dispatchEvent(
  new DialogFailedEvent({
    title: {{ "Failed to Load Video JPEG Quality" | translate }},
    details: error,
  })
);
```

So that's no longer valid JS because Flask would have to pre-render it. We'd have to use different formatters and linters.

And then if we change UI strings, we need to find volunteers to re-translate, and then we're in a bad position if a page is just partially translated.

So it's unfortunately a big cost to the dev workflow, and it's hard for us to accommodate that with the size of TinyPilot's dev team.

**Reply by yeyeto2788** [reply]
>The hard part is rearchitecting our app so that all UI strings go through a translation layer. 

Ohhhh I see what you mean, for sure it is a huge refactor. 


>And then if we change UI strings, we need to find volunteers to re-translate, and then we're in a bad position if a page is just partially translated.

I'm ok with translating regularly with no problem just, just ping me 😋 

Thank you once again for your kind and well explained answers. I hope seeing this project grow much more than what it has all this time. 
Best.
